### PR TITLE
changes: add additional information to properties in GitHubPullrequestPoller

### PR DESCRIPTION
**Pull request by noc0lour**
I want to checkout if this would have chances to get included into the main tree once I'm done refining it. 
Basically I'd like to extract more information from the PR and add it to the properties which then can be used in the builders.

Yes, no? Or in a different way?

### CODE REVIEW

**Comment by noc0lour** [comment]
@tardyp this is more a "design" question I guess, if we allow additional special properties to propagate from different source steps.

**Comment by tardyp** [comment]
You can look at the gerrit change source source code, it has similar code.
It does create one property per event attribute (recursively e.g
{
     author: {'email': 'foo@bar'}
}
will create a property 'gerrit_event.author.email'

**Comment by noc0lour** [comment]
A cool, I'll have a look, probably I want to flatten them as well

**Comment by tardyp** [comment]
indeed, this is a mistake that I made for gerrit design, we should have a whitelist configuration in the changesource (maybe fnmatch based).

**Comment by noc0lour** [comment]
I'd like to have the parsing of PRs in the webhook the same as it is in this PR poller. (e.g. add these changes there as well :) ). Should I just copy & paste ?

**Comment by seankelly** [comment]
Yes. The poller and webhook should stay in sync for what features they support.

File: master/buildbot/changes/github.py [hunk 1]
```diff
@@ -42,9 +43,31 @@
...
+        "github.base.ref",
+        "github.base.sha",
+        "github.base.repo.full_name",
+        "github.base.private",
+        "github.base.fork",
+        "github.merged",
+        "github.mergeable",
+        "github.mergeable_state",
+        "github.merged_by",
+    ]
```
Comment (tardyp) [inline master/buildbot/changes/github.py]: looks like a lot of properties for the defaults.
This should be configurable

File: master/buildbot/changes/github.py [hunk 2]
```diff
@@ -171,6 +198,24 @@ def _getStateObjectId(self):
             '%s/%s' % (self.owner, self.repo), self.db_class_name)
         defer.returnValue(result)
 
+    def _parseGitHubPR(self, pr_info):
+        def flatten(properties, base, info_dict):
+            for k, v in iteritems(info_dict):
+                name = ".".join([base, k])
+                if not any([
+                        prop.startswith(name)
+                        for prop in self.GH_PROPERTY_WHITELIST
```
Comment (tardyp) [inline master/buildbot/changes/github.py]: I would profer is you used fnmatch https://docs.python.org/2/library/fnmatch.html

File: master/buildbot/www/hooks/github.py [hunk 3]
```diff
@@ -108,6 +130,23 @@ def _get_payload(self, request):
 
         return payload
 
+    def _get_PR_information(self, pr_info):
+        def flatten(properties, base, info_dict):
```
Comment (tardyp) [inline master/buildbot/www/hooks/github.py]: I dont like to copy the exact same code in two different places. maybe we should have it in a changes/github.py mixin class

File: master/buildbot/changes/github.py (hunk 2 above)
Comment (noc0lour) [inline master/buildbot/changes/github.py]: I looked into fnmatch but it didn't seem to provide a lot of flexibility if I use a preconfigured list. I'll look into making the list of properties configurable.

**Comment by tardyp** [comment]
I was thinking of a list of fnmatch expr

**Comment by noc0lour** [comment]
I think the drawback of having a `fnmatch` expression is the whole tree has to be flattened first before it can be applied. Otherwise it couldn't match subtrees like `github.base`. or we match backwards (e.g. `fnmatch.match(desired_key, key+"*")` but then we don't have any advantages over using `startswith`.
if you are ok with first flatten the whole file and then matching desired entries (I'd prefer `re` over `fnmatch` anytime).

**Comment by noc0lour** [comment]
I have no idea why the coverage suddenly drops in files I did not even touch? And random things fail in the CI tests? @tardyp ? All files I touched have the same or better coverage.

**Comment by tardyp** [comment]
@noc0lour its because py3.5 tests fail, thus the coverage of py3 is not taken in account

File: master/buildbot/test/unit/test_changes_github.py [hunk 4]
```diff
@@ -162,6 +239,11 @@ def test_SimplePR(self):
         self.assertEqual(change['repository'],
                          'https://github.com/defunkt/buildbot.git')
         self.assertEqual(change['files'], ['README.md'])
+        self.assertDictContainsSubset(_GH_PARSED_PROPS, change['properties'])
```
Comment (tardyp) [inline master/buildbot/test/unit/test_changes_github.py]: this is deprecated in py3.2 (and we dont allow warnings)

https://docs.python.org/3/whatsnew/3.2.html?highlight=assertdictcontainssubset

you probably will need to replace that with a for k,v in _GH_PARSED_PROPS.items()

File: master/buildbot/test/unit/test_changes_github.py [hunk 5]
```diff
@@ -210,6 +297,11 @@ def test_baseURL(self):
...
```
Comment (tardyp) [inline master/buildbot/test/unit/test_changes_github.py]: ditto

File: master/buildbot/changes/github.py [hunk 6]
```diff
@@ -39,10 +41,26 @@
...
 
 
+class PullRequestMixin(object):
+    def extractProperties(self, pr_info):
+        def flatten(properties, base, info_dict):
+            for k, v in iteritems(info_dict):
+                name = ".".join([base, k])
+                if isinstance(v, dict):
+                    flatten(properties, name, v)
+                elif any([fnmatch(name, expr)
```
Comment (tardyp) [inline master/buildbot/changes/github.py]: 👍

**Comment by noc0lour** [comment]
Yay, codecov is happy again :)

**Comment by noc0lour** [comment]
anything left :) ?

**Pull request merged**